### DISCUSSION TITLE
XBuild

### DISCUSSION THREAD

**Discussion by SameerAhmed419** [rust-mobile/xbuild#188] [Q&A]
How can I use XBuild to compile an app made using Xilem into an APK?

**Comment by dvc94ch** [comment]
Don't think xilem requires any special treatment, just build it like you'd build another rust project. If it doesn't build it's because xilem doesn't support android yet. Last I checked xilem didn't have support for images or fonts yet. If you do manage to get something working, I'm sure people would be interested in some screenshots

**Reply by SameerAhmed419** [reply]
Thanks a lot. If I manage to get something working, I will share some screenshots.

**Reply by SameerAhmed419** [reply]
I made a new project using 'cargo new apk-try'. Then, I copied a pasted some code that used Xilem. It ran successfully with 'cargo run'. But, 'x build --target=android --arch=amd64' gave me an apk file with just the AppManifest.xml. Does this mean that Xilem does not support Android?
I want to make an Android app in pure Rust. An incomplete, experimental crate is welcome too. Can you guide me in this regard?
Thanks.

**Reply by dvc94ch** [reply]
I don't think we support amd64 builds. If you connect a real device using a USB cable it should build a valid apk

**Reply by SameerAhmed419** [reply]
I meant arm64.

**Reply by SameerAhmed419** [reply]
Pardon me for the inconvenience. But, could you please provide me with some more detail? I would like to know whether the directory structure has to be like that of the template formed by 'x new [package name]'. Do I have to make a new library crate or a binary crate? What does the crate type have to be? 
Thanks.

**Reply by dvc94ch** [reply]
Don't know, a lot has changed since I worked on xbuild. For Android you always need some java activity which calls rust through the jni. It used to be a Native activity now there are other more popular entry points. You'll need to familiarize yourself with the android startup process and the activity that is the entry point of your app. Don't have time to get into it again. But you'll need a cdylib crate time exposing a C API compatible with that activity.

**Reply by dvc94ch** [reply]
I think we could do better by compiling rust for the JVM. Found a wasm to JVM compiler somewhere, that would be something worth investigating. But this will require some significant effort and you'd need to get into the weeds
